### DISCUSSION TITLE
Distributed training of jax models

### DISCUSSION THREAD

**Discussion by rosikand** [google/flax#2284] [Q&A]
Hi! 

I want to understand how to build, initialize, and train a simple image classifier neural network across 8 TPU cores using a GCP TPU VM or Google Colab TPU. 

For example, how would one adapt the annotated mnist example and parallelize it across multiple cores? The documentation regarding `pmap` seems quite confusing to apply in the context of flax NN's. Is it possible to simply wrap your code in some function call that does the parallelization automatically (like `jax.jit`)? Or does the distribution to each core have to be done manually? Would `pjit` be useful here? 

Lastly, would frameworks such as [ray](https://github.com/ray-project/ray) help?

I am mainly looking for simple, minimal examples for training flax networks across more than one core in a simple manner. 

Thanks!

**Comment by mwitiderrick** [comment]


**Reply by Gonzih** [reply]
paywall, Sadge

**Reply by mwitiderrick** [reply]
https://www.machinelearningnuggets.com/distributed-training-with-jax-and-flax/

**Comment by functionstackx** [comment]
check out this

https://github.com/sholtodouglas/scalingExperiments

**Comment by cgarciae** [comment]
Hey @rosikand, I created this notebook showing how to perform data parallel training with `pmap` + Flax: [MNIST - Data Parallel](https://colab.research.google.com/drive/1hXns2b6T8T393zSrKCSoUktye1YlSe8U#scrollTo=KvuEA8Tw-MYa). It expands the existing MNIST tutorial and also shows how to handle Dropout, BatchNorm, and Metrics in a distributed setting.

@8bitmp3 maybe we can create a guide for `pmap` in addition to `pjit`.
